server/aws-lsp-codewhisperer/src/language-server/localProjectContext/localProjectContextController.ts

```diff
@@ -0,0 +1,203 @@
+                await this._vecLib?.buildIndex(sourceFiles, rootDir, 'all')
+            }
+        } catch (error) {
+            this.log.error(`Error building index: ${error}`)
+        }
+    }
+
+    public async updateWorkspaceFolders(added: WorkspaceFolder[], removed: WorkspaceFolder[]): Promise<void> {
+        try {
+            if (this._vecLib) {
+                const afterRemovals = this.workspaceFolders.filter(
+                    existing => !removed.some(removal => this.areWorkspaceFoldersEqual(existing, removal))
+                )
+
+                const merged = [...afterRemovals]
+                for (const addition of added) {
+                    if (!merged.some(existing => this.areWorkspaceFoldersEqual(existing, addition))) {
+                        merged.push(addition)
+                    }
+                }
+                await this.buildIndex()
+            }
+        } catch (error) {
+            this.log.error(`Error in updateWorkspaceFolders: ${error}`)
+        }
+    }
+
+    public async queryInlineProjectContext(
+        request: QueryInlineProjectContextRequestV2
+    ): Promise<InlineProjectContext[]> {
+        if (!this._vecLib) {
+            return []
+        }
+
+        try {
+            const resp = await this._vecLib?.queryInlineProjectContext(request.query, request.filePath, request.target)
+            return resp ?? []
+        } catch (error) {
+            this.log.error(`Error in queryInlineProjectContext: ${error}`)
+            return []
+        }
+    }
+
+    public async queryVectorIndex(request: QueryRequest): Promise<Chunk[]> {
+        if (!this._vecLib) {
+            return []
+        }
+
+        try {
+            const resp = await this._vecLib?.queryVectorIndex(request.query)
+            return resp ?? []
+        } catch (error) {
+            this.log.error(`Error in queryVectorIndex: ${error}`)
+            return []
+        }
+    }
+
+    private async processWorkspaceFolders(workspaceFolders?: WorkspaceFolder[] | null): Promise<string[]> {
+        const workspaceSourceFiles: string[] = []
+        if (workspaceFolders) {
+            for (const folder of workspaceFolders) {
+                const folderPath = new URL(folder.uri).pathname
+                this.log.info(`Processing workspace: ${folder.name}`)
+
+                try {
+                    const sourceFiles = await this.getCodeSourceFiles(folderPath)
+                    workspaceSourceFiles.push(...sourceFiles)
+                } catch (error) {
+                    this.log.error(`Error processing ${folder.name}: ${error}`)
+                }
+            }
+        }
+        this.log.info(`Found ${workspaceSourceFiles.length} source files`)
+        return workspaceSourceFiles
+    }
+
+    private async getCodeSourceFiles(dir: string): Promise<string[]> {
+        try {
+            const files = await fs.readdir(dir, { withFileTypes: true })
```

Comment:
This will not work in browser. We have `workspace.fs` feature, but it's currently mainly noop for browser. We need to implement reading ws folders for browser.

Reply:
It is not currently intended to work in the browser. This would require significant changes to the underlying library to work within the context of a browser file system and web assemblies. Not in scope for this change - is there anything additionally that needs to be done to exclude this from browser distributions?

Reply:
Yes, excluding it from bundle and handling as optional dependency in other servers should be enough. Some webpack polyfills may be needed in app runtime packages.